rasa_sdk/knowledge_base/actions.py

```diff
@@ -0,0 +1,290 @@
+        elif attribute:
+            return self._query_attribute(dispatcher, tracker)
+
+        dispatcher.utter_template(
+            "Sorry, I did not get that. Can you please rephrase?", tracker
+        )
+        return []
+
+    def _query_entities(self, dispatcher, tracker):
+        """
+        Queries the knowledge base for entities of the requested entity type and
+        outputs those to the user.
+
+        Args:
+            dispatcher: the dispatcher
+            tracker: the tracker
+
+        Returns: list of slots
+        """
+        entity_type = tracker.get_slot(SLOT_ENTITY_TYPE)
+
+        attributes = self._get_attributes_of_entity(entity_type, tracker)
+        entities = self.knowledge_base.get_entities(entity_type, attributes)
+
+        if not entities:
+            dispatcher.utter_template(
+                "I could not find any entities of type '{entity_type}'.", tracker
+            )
+            return []
+
+        representation_function = self.knowledge_base.schema[entity_type][
+            SCHEMA_KEYS_REPRESENTATION
+        ]
+
+        dispatcher.utter_template(
+            "Found the following entities of type '{entity_type}':", tracker
+        )
+        for i, e in enumerate(entities, 1):
+            dispatcher.utter_template(
+                "{}: {}".format(i, representation_function(e)), tracker
+            )
+
+        key_attribute = self.knowledge_base.schema[entity_type][SCHEMA_KEYS_KEY]
+
+        last_entity = None if len(entities) > 1 else entities[0][key_attribute]
+
+        slots = [
+            SlotSet(SLOT_ENTITY_TYPE, entity_type),
+            SlotSet(SLOT_MENTION, None),
+            SlotSet(SLOT_ATTRIBUTE, None),
+            SlotSet(SLOT_LAST_ENTITY, last_entity),
+            SlotSet(SLOT_LAST_ENTITY_TYPE, entity_type),
+            SlotSet(SLOT_LISTED_ITEMS, list(map(lambda e: e[key_attribute], entities))),
+        ]
+
+        return slots + self._reset_attribute_slots(entity_type, tracker)
+
+    def _query_attribute(self, dispatcher, tracker):
+        """
+        Queries the knowledge base for the value of the requested attribute of the
+        mentioned entity and outputs it to the user.
+
+        Args:
+            dispatcher: the dispatcher
+            tracker: the tracker
+
+        Returns: list of slots
+        """
+        entity_type = tracker.get_slot(SLOT_ENTITY_TYPE)
+        attribute = tracker.get_slot(SLOT_ATTRIBUTE)
+
+        entity = self._get_entity(tracker)
+
+        if not entity or not attribute:
+            dispatcher.utter_template(
+                "Sorry, I did not get that. Can you please rephrase?", tracker
+            )
+
+            slots = [SlotSet(SLOT_MENTION, None)]
+            return slots + self._reset_attribute_slots(entity_type, tracker)
+
+        value = self.knowledge_base.get_attribute_of(entity_type, entity, attribute)
+
+        # utter response
+        if value:
```

Comment:
maybe also wrap this in a function that's overridable so people can change the text without copying the logic in this function